From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 29 Jun 2025 09:35:02 +0930
Subject: [PATCH] Dispatch Discord commands through CommandMap


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 716c3b17dc07eb2417dfb0084fe75dbd270f031a..6696d4dfcdf9ce1a2dec24195f478cfeb8196b3f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -917,6 +917,11 @@ public final class CraftServer implements Server {
         Preconditions.checkArgument(rawSender != null, "sender cannot be null");
         Preconditions.checkArgument(commandLine != null, "commandLine cannot be null");
         org.spigotmc.AsyncCatcher.catchOp("Command Dispatched Async: " + commandLine); // Spigot // Paper - Include command in error message
+        // Nabulus start
+        if (rawSender instanceof org.bukkit.OfflinePlayer && rawSender instanceof org.bukkit.command.ConsoleCommandSender) {
+            return this.commandMap.dispatch(rawSender, commandLine);
+        }
+        // Nabulus end
         CommandSourceStack sourceStack = VanillaCommandWrapper.getListener(rawSender);
 
         String command = StringUtils.normalizeSpace(commandLine.trim());
