From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 29 Jun 2025 09:35:02 +0930
Subject: [PATCH] Dispatch Discord commands through CommandMap


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 478f4461ab3fb21b288aa2f6079860f23832e622..7683e4be8af234ddb0cd5b07014b2a1f80508cec 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -920,6 +920,11 @@ public final class CraftServer implements Server {
         Preconditions.checkArgument(rawSender != null, "sender cannot be null");
         Preconditions.checkArgument(commandLine != null, "commandLine cannot be null");
         org.spigotmc.AsyncCatcher.catchOp("Command Dispatched Async: " + commandLine); // Spigot // Paper - Include command in error message
+        // Nabulus start
+        if (rawSender instanceof org.bukkit.OfflinePlayer && rawSender instanceof org.bukkit.command.ConsoleCommandSender) {
+            return this.commandMap.dispatch(rawSender, commandLine);
+        }
+        // Nabulus end
         CommandSourceStack sourceStack = VanillaCommandWrapper.getListener(rawSender);
 
         String command = StringUtils.normalizeSpace(commandLine.trim());
