From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Sun, 27 Feb 2022 21:27:01 +0930
Subject: [PATCH] Separate creative levels


diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V1022.java b/ca/spottedleaf/dataconverter/minecraft/versions/V1022.java
index 0c7386ebc9346aa7fd2220b05a12f9538f1400cc..5268efee271bd96cc74a9a7e1f6fc1530e20f92b 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V1022.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V1022.java
@@ -20,6 +20,7 @@ public final class V1022 {
 
                 WalkerUtils.convertList(MCTypeRegistry.ITEM_STACK, data, "Inventory", fromVersion, toVersion);
                 WalkerUtils.convertList(MCTypeRegistry.ITEM_STACK, data, "EnderItems", fromVersion, toVersion);
+            WalkerUtils.convertList(MCTypeRegistry.ITEM_STACK, data, "Nabulus.CreativeLevelInventory", fromVersion, toVersion);
 
                 WalkerUtils.convert(MCTypeRegistry.ENTITY, data, "ShoulderEntityLeft", fromVersion, toVersion);
                 WalkerUtils.convert(MCTypeRegistry.ENTITY, data, "ShoulderEntityRight", fromVersion, toVersion);
diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V4312.java b/ca/spottedleaf/dataconverter/minecraft/versions/V4312.java
index cc39b5186b511f99f5c8fc212cae96d3c2960db9..9fcb373cfa44ecc428194313b5fbd706a51f0f5b 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V4312.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V4312.java
@@ -28,7 +28,14 @@ public final class V4312 {
 
             @Override
             public MapType convert(final MapType data, final long sourceVersion, final long toVersion) {
-                final ListType inventory = data.getListUnchecked("Inventory");
+                // Nabulus start
+                convertInventory(data, "Inventory", "equipment");
+                convertInventory(data, "Nabulus.CreativeLevelInventory", "Nabulus.CreativeLevelEquipment");
+                return null;
+            }
+            private Void convertInventory(final MapType data, String inventoryDataKey, String equipmentDataKey) {
+                final ListType inventory = data.getListUnchecked(inventoryDataKey);
+                // Nabulus end
                 if (inventory == null) {
                     return null;
                 }
@@ -54,7 +61,7 @@ public final class V4312 {
                     equipment.setMap(equipmentKey, item);
                 }
 
-                data.setMap("equipment", equipment);
+                data.setMap(equipmentDataKey, equipment); // Nabulus
 
                 return null;
             }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 23b3a82ab3e1c6abc85000d9285470cb12300a6b..b2afb34d4a295f0f405210e2470f7c4c1fdd0ac0 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1542,6 +1542,13 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             this.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.NO_RESPAWN_BLOCK_AVAILABLE, 0.0F));
         }
 
+        // Nabulus start
+        boolean toOrFromCreativeLevel = transition.newLevel().getWorld().isCreativeLevel() || this.level().getWorld().isCreativeLevel();
+        if (toOrFromCreativeLevel) {
+            this.releaseLeftShoulderEntity();
+            this.releaseRightShoulderEntity();
+        }
+        // Nabulus end
         ServerLevel newLevel = transition.newLevel();
         ServerLevel oldLevel = this.level();
         ResourceKey<Level> lastDimension = oldLevel.dimension();
@@ -1598,6 +1605,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         profiler.pop();
         profiler.push("placing");
         this.setServerLevel(newLevel);
+        this.chooseInventory(); // Nabulus
         this.connection.internalTeleport(PositionMoveRotation.of(transition), transition.relatives()); // CraftBukkit - use internal teleport without event
         this.connection.resetPosition();
         newLevel.addDuringTeleport(this);
@@ -2213,11 +2221,21 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     public void restoreFrom(final ServerPlayer oldPlayer, final boolean restoreAll) {
+        // Nabulus start
+        this.restoreFrom(oldPlayer, restoreAll, false);
+    }
+    public void restoreFrom(final ServerPlayer oldPlayer, final boolean restoreAll, boolean toOrFromCreativeLevel) {
+        // Nabulus end
         this.wardenSpawnTracker = oldPlayer.wardenSpawnTracker;
         this.chatSession = oldPlayer.chatSession;
         this.gameMode.setGameModeForPlayer(oldPlayer.gameMode.getGameModeForPlayer(), oldPlayer.gameMode.getPreviousGameModeForPlayer());
         this.onUpdateAbilities();
         this.getAttributes().assignBaseValues(oldPlayer.getAttributes());
+        // Nabulus start
+        this.normalInventory.replaceWith(oldPlayer.normalInventory);
+        this.creativeLevelInventory.replaceWith(oldPlayer.creativeLevelInventory);
+        this.usingNormalInventory = oldPlayer.usingNormalInventory;
+        // Nabulus end
         if (restoreAll) {
             // this.getAttributes().assignPermanentModifiers(oldPlayer.getAttributes()); // CraftBukkit
             this.setHealth(oldPlayer.getHealth());
@@ -2248,6 +2266,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.currentExplosionImpactPos = oldPlayer.currentExplosionImpactPos;
         this.currentExplosionCause = oldPlayer.currentExplosionCause;
         this.chunkTrackingView = oldPlayer.chunkTrackingView;
+        if (toOrFromCreativeLevel) return; // Nabulus
         this.requestedDebugSubscriptions = oldPlayer.requestedDebugSubscriptions;
         this.setShoulderEntityLeft(oldPlayer.getShoulderEntityLeft());
         this.setShoulderEntityRight(oldPlayer.getShoulderEntityRight());
@@ -2766,6 +2785,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public void setServerLevel(final ServerLevel level) {
         this.setLevel(level);
         this.gameMode.setLevel(level);
+        this.chooseInventory(); // Nabulus
     }
 
     private static @Nullable GameType readPlayerMode(final ValueInput playerInput, final String modeTag) {
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index a7f589970a04e520eb5cbe6dfeb333bc9cd479a8..69cd7603ba57690839cf1d269448f930d4145d14 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -609,8 +609,15 @@ public abstract class PlayerList {
         serverPlayer.level().removePlayerImmediately(serverPlayer, removalReason);
         ServerLevel level = respawnInfo.newLevel();
         ServerPlayer player = serverPlayer; // Paper - TODO - recreate instance
+        // Nabulus start
+        boolean toOrFromCreativeLevel = fromLevel.getWorld().isCreativeLevel() || level.getWorld().isCreativeLevel();
+        if (toOrFromCreativeLevel) {
+            player.releaseLeftShoulderEntity();
+            player.releaseRightShoulderEntity();
+        }
+        // Nabulus end
         player.connection = serverPlayer.connection;
-        player.restoreFrom(serverPlayer, keepAllPlayerData);
+        player.restoreFrom(serverPlayer, keepAllPlayerData, toOrFromCreativeLevel); // Nabulus
         player.setId(serverPlayer.getId());
         player.setMainArm(serverPlayer.getMainArm());
         if (false && !respawnInfo.missingRespawnBlock()) { // Paper - Once we not reuse the player entity, this can be flipped again but without the events being fired
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index cf3619b6de23f32e9e5a81fbe287fb438ecfcabc..c334e04c1f636211535d462e82ffb67a1b227496 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4207,6 +4207,11 @@ public abstract class Entity
         }
         // Paper end - Fix item duplication and teleport issues
         if (this.level() instanceof ServerLevel serverLevel && !this.isRemoved()) {
+            // Nabulus start
+            if (!this.level.equals(serverLevel) && this.level().getWorld().isCreativeLevel()) {
+                return null;
+            }
+            // Nabulus end
             // CraftBukkit start
             PositionMoveRotation absolutePosition = PositionMoveRotation.calculateAbsolute(PositionMoveRotation.of(this), PositionMoveRotation.of(transition), transition.relatives());
             Vec3 velocity = absolutePosition.deltaMovement(); // Paper
diff --git a/net/minecraft/world/entity/PortalProcessor.java b/net/minecraft/world/entity/PortalProcessor.java
index 7ec15f3d3a770e1e9e2fbac29dd38e947bb07091..375610a8a544926f63bccd8434f94c895b655786 100644
--- a/net/minecraft/world/entity/PortalProcessor.java
+++ b/net/minecraft/world/entity/PortalProcessor.java
@@ -29,6 +29,11 @@ public class PortalProcessor {
     }
 
     public @Nullable TeleportTransition getPortalDestination(final ServerLevel serverLevel, final Entity entity) {
+        // Nabulus start
+        if (entity.level().getWorld().isCreativeLevel()) {
+            return null;
+        }
+        // Nabulus end
         return this.portal.getPortalDestination(serverLevel, entity, this.entryPosition);
     }
 
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 4068dc1cea4edca196dc0c58b4804e94dd07881e..fec8c62191014c5df073fa7c102b93c4eb802ef7 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -179,6 +179,36 @@ public abstract class Player extends Avatar implements ContainerUser {
     public boolean fauxSleeping;
     public int oldLevel = -1;
 
+    // Nabulus start
+    protected final Inventory normalInventory = new Inventory(this, this.createEquipment());
+    protected final Inventory creativeLevelInventory = new Inventory(this, this.createEquipment());
+    public boolean usingNormalInventory = true;
+
+    public void chooseInventory() {
+        if (this.level().getWorld().isCreativeLevel()) {
+            if (usingNormalInventory) {
+                this.normalInventory.replaceWith(this.inventory);
+                this.inventory.replaceWith(this.creativeLevelInventory);
+                usingNormalInventory = false;
+            }
+        } else {
+            if (!usingNormalInventory) {
+                this.creativeLevelInventory.replaceWith(this.inventory);
+                this.inventory.replaceWith(this.normalInventory);
+                usingNormalInventory = true;
+            }
+        }
+    }
+
+    public Inventory getNormalInventory() {
+        return usingNormalInventory ? this.inventory : this.normalInventory;
+    }
+
+    public Inventory getCreativeLevelInventory() {
+        return usingNormalInventory ? this.creativeLevelInventory : this.inventory;
+    }
+    // Nabulus end
+
     @Override
     public org.bukkit.craftbukkit.entity.CraftHumanEntity getBukkitEntity() {
         return (org.bukkit.craftbukkit.entity.CraftHumanEntity) super.getBukkitEntity();
@@ -656,6 +686,14 @@ public abstract class Player extends Avatar implements ContainerUser {
         this.setUUID(this.gameProfile.id());
         this.inventory.load(input.listOrEmpty("Inventory", ItemStackWithSlot.CODEC));
         this.inventory.setSelectedSlot(input.getIntOr("SelectedItemSlot", 0));
+        // Nabulus start
+        this.normalInventory.replaceWith(this.inventory);
+        this.creativeLevelInventory.load(input.listOrEmpty("Nabulus.CreativeLevelInventory", ItemStackWithSlot.CODEC));
+        this.creativeLevelInventory.setSelectedSlot(input.getIntOr("Nabulus.CreativeLevelSelectedItemSlot", 0));
+        this.creativeLevelInventory.equipment.setAll(input.read("Nabulus.CreativeLevelEquipment", EntityEquipment.CODEC).orElseGet(EntityEquipment::new));
+        this.usingNormalInventory = true;
+        chooseInventory();
+        // Nabulus end
         this.sleepCounter = input.getShortOr("SleepTimer", (short)0);
         this.experienceProgress = input.getFloatOr("XpP", 0.0F);
         this.experienceLevel = input.getIntOr("XpLevel", 0);
@@ -675,10 +713,24 @@ public abstract class Player extends Avatar implements ContainerUser {
 
     @Override
     protected void addAdditionalSaveData(final ValueOutput output) {
+        // Nabulus start
+        if (!usingNormalInventory) {
+            this.creativeLevelInventory.replaceWith(this.inventory);
+            this.inventory.replaceWith(this.normalInventory);
+        }
+        // Nabulus end
         super.addAdditionalSaveData(output);
         NbtUtils.addCurrentDataVersion(output);
         this.inventory.save(output.list("Inventory", ItemStackWithSlot.CODEC));
         output.putInt("SelectedItemSlot", this.inventory.getSelectedSlot());
+        // Nabulus start
+        this.creativeLevelInventory.save(output.list("Nabulus.CreativeLevelInventory", ItemStackWithSlot.CODEC));
+        output.putInt("Nabulus.CreativeLevelSelectedItemSlot", this.creativeLevelInventory.getSelectedSlot());
+        output.store("Nabulus.CreativeLevelEquipment", EntityEquipment.CODEC, this.creativeLevelInventory.equipment);
+        if (!usingNormalInventory) {
+            this.inventory.replaceWith(this.creativeLevelInventory);
+        }
+        // Nabulus end
         output.putShort("SleepTimer", (short)this.sleepCounter);
         output.putFloat("XpP", this.experienceProgress);
         output.putInt("XpLevel", this.experienceLevel);
