From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: froobynooby <froobynooby@froobworld.com>
Date: Fri, 10 Feb 2023 19:54:12 +0930
Subject: [PATCH] Add silent chest flag to players


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 3a4da514c163e7565fc241835db52c464a5c4734..27375794d85539cfe989d82e26d63e45a079e6cc 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -306,6 +306,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper - Friction API
     public int invulnerableDuration = LivingEntity.INVULNERABLE_DURATION; // Paper - configurable invulnerable duration
     // CraftBukkit end
+    public boolean silentChests = false; // Nabulus
 
     protected LivingEntity(final EntityType<? extends LivingEntity> type, final Level level) {
         super(type, level);
diff --git a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
index 52737172ccfa0c7c977e4f2fe1db242e0bb92d25..1fed9723381081b1de42ae1be11596d7840fa98c 100644
--- a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
+++ b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
@@ -48,6 +48,7 @@ public abstract class ContainerOpenersCounter {
     public void incrementOpeners(
         final LivingEntity entity, final Level level, final BlockPos pos, final BlockState blockState, final double maxInteractionRange
     ) {
+        if (entity.silentChests) return; // Nabulus
         // Paper start - delay open/close callbacks
         if (this.delayCallbacks()) {
             scheduleRecheck(level, pos, blockState, 1);
@@ -78,6 +79,7 @@ public abstract class ContainerOpenersCounter {
     }
 
     public void decrementOpeners(final LivingEntity entity, final Level level, final BlockPos pos, final BlockState blockState) {
+        if (entity.silentChests) return; // Nabulus
         // Paper start - delay open/close callbacks
         if (this.delayCallbacks()) {
             scheduleRecheck(level, pos, blockState, 1);
@@ -110,7 +112,7 @@ public abstract class ContainerOpenersCounter {
     public List<ContainerUser> getEntitiesWithContainerOpen(final Level level, final BlockPos pos) {
         double range = this.maxInteractionRange + 4.0;
         AABB searchBox = new AABB(pos).inflate(range);
-        return level.getEntities((Entity)null, searchBox, entity -> this.hasContainerOpen(entity, pos))
+        return level.getEntities((Entity)null, searchBox, entity -> this.hasContainerOpen(entity, pos) && (!(entity instanceof Player) || !((Player) entity).silentChests)) // Nabulus - add silent chest check
             .stream()
             .map(entity -> (ContainerUser)entity)
             .collect(Collectors.toList());
